### PR TITLE
TASK: Workspace aware commands events migration

### PR DESCRIPTION
**Pull request by mhsdesign**
See https://github.com/neos/neos-development-collection/pull/4708

### CODE REVIEW

File: Neos.ContentRepositoryRegistry/Classes/Command/MigrateEventsCommandController.php [hunk 1]
```diff
@@ -32,4 +32,20 @@ public function migratePropertiesToUnsetCommand(string $contentRepository = 'def
         $eventMigrationService = $this->contentRepositoryRegistry->buildService($contentRepositoryId, $this->eventMigrationServiceFactory);
         $eventMigrationService->migratePropertiesToUnset($this->outputLine(...));
     }
+
+    /**
+     * Adds a dummy workspace name to the events meta-data, so it can be replayed.
+     *
+     * Needed for #4708: https://github.com/neos/neos-development-collection/pull/4708
+     *
+     * Included in February 2023 - before final Neos 9.0 release
```
Comment (ahaeslich) [inline Neos.ContentRepositoryRegistry/Classes/Command/MigrateEventsCommandController.php]:
```suggestion
     * Included in March 2024 - before final Neos 9.0 release
```

**Review by nezaniel — APPROVED** [review]
Nicely done

**Review by ahaeslich — CHANGES_REQUESTED** [review]
Tested it locally and run into this error while replaying the contentStream projection:

```
Undefined array key "workspaceName" in /application/Packages/Neos/Neos.ContentRepository.Core/Classes/Feature/WorkspacePublication/Dto/NodeIdToPublishOrDiscard.php line 46
```

**Review by ahaeslich — APPROVED** [review]

**Comment by mhsdesign** [comment]
@ahaeslich looks perfect thanks, we will merge this tomorrow once we have @bwaidelich conceptional approval.

File: Neos.ContentRepository.Core/Classes/Feature/Common/ContentStreamIdOverride.php [hunk 2]
```diff
@@ -24,13 +27,41 @@ class ContentStreamIdOverride
...
+            self::$contentStreamIdToUse = null;
+            throw $th;
+        }
+        self::$contentStreamIdToUse = null;
+    }
+
+    /**
+     * @internal
+     */
+    public static function findContentStreamIdForWorkspace(ContentRepository $contentRepository, WorkspaceName $workspaceName): ContentStreamId
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/Feature/Common/ContentStreamIdOverride.php]: nitpick: We have the guideline to use `find*` for methods that can return `null`
```suggestion
    public static function getContentStreamIdForWorkspace(ContentRepository $contentRepository, WorkspaceName $workspaceName): ContentStreamId
```

File: Neos.ContentRepository.Core/Classes/Feature/Common/ContentStreamIdOverride.php [hunk 3]
```diff
@@ -24,13 +27,41 @@ class ContentStreamIdOverride
...
      * A content stream id that to use instead of the workspace one's {@see ConstraintChecks::requireContentStream()}
      */
-    public static ?ContentStreamId $contentStreamIdToUse = null;
+    private static ?ContentStreamId $contentStreamIdToUse = null;
 
     /**
      * @internal
      */
-    public static function useContentStreamId(?ContentStreamId $contentStreamIdToUse): void
+    public static function withContentStreamIdToUse(ContentStreamId $contentStreamIdToUse, \Closure $fn): void
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/Feature/Common/ContentStreamIdOverride.php]: "withContentStreamToUse" is a bit bumpy.. I can't think of something better atm.
But more importantly we should add a doc comment explaining this unusual construct even though it is just used internally

File: Neos.ContentRepository.Core/Classes/Feature/Common/ContentStreamIdOverride.php [hunk 4]
```diff
@@ -24,13 +27,41 @@ class ContentStreamIdOverride
...
+        if (self::$contentStreamIdToUse !== null) {
+            throw new \Exception('Recursive content stream override is not supported', 1710426945);
+        }
         self::$contentStreamIdToUse = $contentStreamIdToUse;
+        try {
+            $fn();
+        } catch (\Throwable $th) {
+            self::$contentStreamIdToUse = null;
+            throw $th;
+        }
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/Feature/Common/ContentStreamIdOverride.php]:
```suggestion
        } finally {
            self::$contentStreamIdToUse = null;
        }
```

File: Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php [hunk 5]
```diff
@@ -222,6 +231,88 @@ public function migratePropertiesToUnset(\Closure $outputFn)
...
+     *
+     * @param \Closure $outputFn
+     * @return void
+     */
+    public function migrateMetaDataToWorkspaceName(\Closure $outputFn)
+    {
+        $this->eventsModified = [];
+
+        $backupEventTableName = DoctrineEventStoreFactory::databaseTableName($this->contentRepositoryId)
+            . '_bak_' . date('Y_m_d_H_i_s');
```
Comment (bwaidelich) [inline Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php]: another nitpicky suggestion:
```suggestion
            . '_bkp_' . date('Y_m_d_H_i_s');
```

**Review by bwaidelich — COMMENTED** [review]
Looks great already, just some (partly quite nitpicky) suggestions

**Comment by ahaeslich** [comment]
For documenation: replaying all projects now worked on the projects database.

File: Neos.ContentRepository.Core/Classes/Feature/Common/ContentStreamIdOverride.php [hunk 6]
```diff
@@ -40,9 +40,8 @@ public static function withContentStreamIdToUse(ContentStreamId $contentStreamId
         self::$contentStreamIdToUse = $contentStreamIdToUse;
         try {
             $fn();
-        } catch (\Throwable $th) {
+        } finally {
             self::$contentStreamIdToUse = null;
-            throw $th;
         }
         self::$contentStreamIdToUse = null;
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/Feature/Common/ContentStreamIdOverride.php]: this line can now be removed, sorry forgot that

File: Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php (hunk 5 above)
Comment (mhsdesign) [reply Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php]: followed the pattern of another migration, please adjust also there if here ^^ // cc @ahaeslich

File: Neos.ContentRepository.Core/Classes/Feature/Common/ContentStreamIdOverride.php (hunk 3 above)
Comment (kitsunet) [reply Neos.ContentRepository.Core/Classes/Feature/Common/ContentStreamIdOverride.php]: I would actually avoid `with` as it doesn't follow our usual pattern of how `with` work (aka return a new instance with the property applied).
how about `ContentStreamIdOverride::applyContentStreamId` or even `applyContentStreamIdToClosure` to be very specific.

Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/Feature/Common/ContentStreamIdOverride.php]: yeah sounds fine the `with` pattern is actually overloaded if you think of the `withoutAuthentication` function which also accepts a closure.

I sadly havent found yet a good prefix for these kinds of functions that set state and cleanup while another function is called.

**Review by ahaeslich — APPROVED** [review]

**Pull request merged**